### ISSUE TITLE
LineReceiver rejects lines of length MAX_LENGTH sometimes

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6556]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)| @zooko reported|
|-|-|
|Trac ID|trac#6556|
|Type|enhancement|
|Created|2013-06-08 20:41:38Z|
|Branch|https://github.com/twisted/twisted/tree/line-receiver-max-length-fragmentation-6556|

If ```MAX_LENGTH``` is 3 and ```delimiter``` is '\r\n', and I send "xyz\r" and then "\n" in two packets, ```LineReceiver``` will disconnect after receiving the first packet, calling ```lineLengthExceeded```. On the other hand if I send "xyz\r\n" in one packet, ```LineReceiver``` will accept the line.

Here is a test. I propose a fix: ```LineReceiver``` should not check for line-length-exceeded until the number of bytes of the data >= ```MAX_LENGTH``` + ```len(delimiter)```.

Attachments:

* [notQuiteMaximumLineLengthUnfinished.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/283/28311f41b614798b9a42a4bea40f02785a6b0d5d/8a6483adbfcb6ea3b231bc08d12bb1eac2494eda.patch) (1229 bytes) - added by zooko on 2013-06-08 20:41:55Z - 
* [test-max-length.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/283/28311f41b614798b9a42a4bea40f02785a6b0d5d/729a9a5731ad24512e368f109cc2641a24ad539e.patch) (21295 bytes) - added by zooko on 2013-06-09 09:29:05Z - 
* [fix-max-length.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/283/28311f41b614798b9a42a4bea40f02785a6b0d5d/e82b8dd5facce3b5ac7c968793a44440e252c65c.patch) (22110 bytes) - added by zooko on 2013-06-11 07:08:28Z - 
* [fix-max-length.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/283/28311f41b614798b9a42a4bea40f02785a6b0d5d/1bfab9511f17a4b6e54e99c95448476532935d5b.patch) (3643 bytes) - added by zooko on 2013-06-11 08:55:35Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__6556 6556
type__enhancement enhancement
reporter__zooko zooko
priority__normal normal
milestone__ 
branch__branches_line_receiver_max_length_fragmentation_6556 branches/line-receiver-max-length-fragmentation-6556
branch_author__tomprince tomprince
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1370724098000000 1370724098000000
changetime__1510260781303678 1510260781303678
version__None None
owner__zooko zooko
cc__zooko@...
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_1"></a>|@zooko commented|
|-|-|

[[test-max-length.patch](None/test-max-length.patch) Here] are some more tests which I developed while working on optimizations for ```LineReceiver``` ([#6357](https://github.com/twisted/twisted/issues/6357)). Some of these tests detected bugs in some of my attempted optimizations, so I consider them useful. Look out for the one disabled, and accompanied by a comment saying "I propose we don't require this behavior.". I'm including it in this patch to solicit opinions from others — do you agree that behavior shouldn't be required?

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_3"></a>|@zooko commented|
|-|-|

[[fix-max-length.2.patch](None/fix-max-length.2.patch) Here] is a patch with just the test and the fix for just this issue. Please review!

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_4"></a>|@zooko commented|
|-|-|

Please see my comment over on comment:4:ticket:6536, which is that I have a working patch for [#6357](https://github.com/twisted/twisted/issues/6357) and am blocked because I would like for that optimization patch to be applied separately, on top of these bugfix patches: [#6556](https://github.com/twisted/twisted/issues/6556), [#6537](https://github.com/twisted/twisted/issues/6537), [#6558](https://github.com/twisted/twisted/issues/6558).

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_5"></a>|@zooko commented|
|-|-|

Please see my comment over on comment:4:ticket:6536, which is that I have a working patch for [#6357](https://github.com/twisted/twisted/issues/6357) and am blocked because I would like for that optimization patch to be applied separately, on top of these bugfix patches: [#6556](https://github.com/twisted/twisted/issues/6556), [#6537](https://github.com/twisted/twisted/issues/6537), [#6558](https://github.com/twisted/twisted/issues/6558).

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_6"></a>|@tomprince commented|
|-|-|

(In [39148]) Branching to line-receiver-max-length-fragmentation-6556.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_7"></a>|@tomprince commented|
|-|-|

(In [39149]) Apply fix-max-length.2.patch from zooko.

Refs: [#6556](https://github.com/twisted/twisted/issues/6556)

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_8"></a>|@tomprince commented|
|-|-|

I guess this will behave pathologically, when the delimiter is pathologically long. If it is a significant fraction of `MAX_LENGTH`, then `MAX_LENGTH + len(delimiter)` will be read, even if none of the characters after `MAX_LENGTH` match the delimiter. I dont think it worth worrying about this though.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_9"></a>|@tomprince set owner to @zooko|
|-|-|

1.There is a python3 failure. It appears due to setting the delimiter explicitly to a string (rather than bytestring).
 2. There are some whitespace and line-length issues reported by [twistedchecker](https://buildbot.twistedmatrix.com/builders/twistedchecker/builds/982/steps/run-twistedchecker/logs/new%20twistedchecker%20errors). 
   - Ignore the error about [https://bugs.launchpad.net/twistedchecker/+bug/1200696 `MAX_LENGTH`] and [comments https://bugs.launchpad.net/twistedchecker/+bug/1200695], as they are bugs in twistedchecker. 
 3. The following assertion won't provide a good error, in the case `proto.received` is `[]`. Please split it into an assertion about `len(proto.received)` and `proto.received[0]`. 
```python
self.assertEqual(line, proto.received and proto.received[0])
```
 4. It appears there isn't any test of receiving a line as long as or longer than `MAX_LENGTH + len(delimter)` without a delimiter.
   - Please add one or more.
 5. Is there any reason you can't use `LineTester`? The extra stuff that it does doesn't appear to affect your tests.
   - Please switch to using it, or explain why it isn't sufficient.

Please resubmit for review, after addressing the above points.

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_10"></a>|@markrwilliams commented|
|-|-|

Replying to [Tom Prince](#note_9):
>  1. There is a python3 failure. It appears due to setting the delimiter explicitly to a string (rather than bytestring).
>  2. There are some whitespace and line-length issues reported by [twistedchecker](https://buildbot.twistedmatrix.com/builders/twistedchecker/builds/982/steps/run-twistedchecker/logs/new%20twistedchecker%20errors). 
>    - Ignore the error about [https://bugs.launchpad.net/twistedchecker/+bug/1200696 `MAX_LENGTH`] and [comments https://bugs.launchpad.net/twistedchecker/+bug/1200695], as they are bugs in twistedchecker. 
>  3. The following assertion won't provide a good error, in the case `proto.received` is `[]`. Please split it into an assertion about `len(proto.received)` and `proto.received[0]`. 
> ```
> #!python
> self.assertEqual(line, proto.received and proto.received[0])
> ```
>  4. It appears there isn't any test of receiving a line as long as or longer than `MAX_LENGTH + len(delimter)` without a delimiter.
>    - Please add one or more.
>  5. Is there any reason you can't use `LineTester`? The extra stuff that it does doesn't appear to affect your tests.
>    - Please switch to using it, or explain why it isn't sufficient.
> 
> Please resubmit for review, after addressing the above points.

https://github.com/twisted/twisted/pull/917

I believe I have addressed all these points.

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_12"></a>|@markrwilliams set status to `closed`|
|-|-|

In changeset 07d861149b33978a74f5e0f706b14fd00968c979
```
#!CommitTicketReference repository="" revision="07d861149b33978a74f5e0f706b14fd00968c979"
Merge pull request #917 from twisted/line-receiver-max-length-fragmentation-6556

Author: zooko, markrwilliams
Reviewer: tomprince, moshez
Fixes: ticket:6556

twisted.protocols.basic.LineReceiver checks received lines' lengths against its MAX_LENGTH only after receiving a complete delimiter.
```
